File: onux/packages/ios-swift/Sources/OnChainUX/ChainAdapter/EVMAdapter.swift
/**
 * EVMChainAdapter — EVM chain adapter for iOS.
 *
 * Provides EVM-specific operations: balance queries, contract calls,
 * gas estimation, and transaction signing through a connected wallet.
 *
 * ## Usage
 * ```swift
 * let adapter = EVMChainAdapter()
 * adapter.setRpcUrl("https://eth.llamarpc.com")
 *
 * let balance = try await adapter.getBalance(address: "0x...")
 * let gasPrice = try await adapter.getGasPrice()
 * let receipt = try await adapter.getTransaction(hash: "0x...")
 * ```
 */

import Foundation

/// Parameters for an EVM contract call.
public struct EthCallParams: Sendable {
    /// Target contract address.
    public let to: String
    /// Encoded function call data (hex).
    public let data: String?
    /// Caller address.
    public let from: String?
    /// ETH value to send (hex).
    public let value: String?
    /// Block number (default: "latest").
    public let blockNumber: String?
    
    public init(
        to: String,
        data: String? = nil,
        from: String? = nil,
        value: String? = nil,
        blockNumber: String? = nil
    ) {
        self.to = to
        self.data = data
        self.from = from
        self.value = value
        self.blockNumber = blockNumber
    }
}

/// EVM chain adapter providing chain-specific operations.
public final class EVMChainAdapter {
    
    /// RPC endpoint URL.
    private var rpcUrl: String = ""
    
    /// Registered chains.
    private var chains: [ChainConfig] = []
    
    /// Create a new EVM adapter.
    public init() {}
    
    // MARK: - Configuration
    
    /// Register supported EVM chains.
    /// - Parameter chains: Array of chain configurations.
    public func registerChains(_ chains: [ChainConfig]) {
        self.chains = chains
    }
    
    /// Set the RPC endpoint URL.
    /// - Parameter url: RPC URL.
    public func setRpcUrl(_ url: String) {
        self.rpcUrl = url
    }
    
    // MARK: - Balance
    
    /// Get the native balance for an address.
    /// - Parameter address: Ethereum address.
    /// - Returns: Balance in wei (hex string).
    public func getBalance(address: String) async throws -> String {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_getBalance", params: [address, "latest"])
        return response as? String ?? "0x0"
    }
    
    /// Get formatted balance in ETH.
    /// - Parameter address: Ethereum address.
    /// - Returns: Balance as a decimal string.
    public func getBalanceFormatted(address: String) async throws -> String {
        let hexBalance = try await getBalance(address: address)
        return hexToEth(hexBalance)
    }
    
    // MARK: - Contract Calls
    
    /// Call a contract read method.
    /// - Parameter params: Call parameters.
    /// - Returns: Result data (hex string).
    public func call(params: EthCallParams) async throws -> String {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let callParams: [String: Any] = buildCallParams(params)
        let blockTag = params.blockNumber ?? "latest"
        let response = try await rpcCall(method: "eth_call", params: [callParams, blockTag])
        return response as? String ?? "0x"
    }
    
    /// Get ERC-20 token balance for an address.
    /// - Parameters:
    ///   - tokenAddress: ERC-20 contract address.
    ///   - userAddress: User wallet address.
    /// - Returns: Token balance (hex string).
    public func getTokenBalance(tokenAddress: String, userAddress: String) async throws -> String {
        // ERC-20 balanceOf function selector + padded address
        let paddedAddress = userAddress
            .replacingOccurrences(of: "0x", with: "")
            .padding(toLength: 64, withPad: "0", startingAt: 0)
        let data = "0x70a08231\(paddedAddress.lowercased())"
        
        return try await call(params: EthCallParams(
            to: tokenAddress,
            data: data
        ))
    }
    
    // MARK: - Gas
    
    /// Estimate gas for a transaction.
    /// - Parameter tx: Transaction parameters.
    /// - Returns: Gas estimate (hex string).
    public func estimateGas(tx: [String: String]) async throws -> String {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_estimateGas", params: [tx])
        return response as? String ?? "0x0"
    }
    
    /// Get the current gas price.
    /// - Returns: Gas price in wei (hex string).
    public func getGasPrice() async throws -> String {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_gasPrice")
        return response as? String ?? "0x0"
    }
    
    // MARK: - Transactions
    
    /// Get a transaction by hash.
    /// - Parameter hash: Transaction hash.
    /// - Returns: Transaction data.
    public func getTransaction(hash: String) async throws -> [String: Any] {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_getTransactionByHash", params: [hash])
        return response as? [String: Any] ?? [:]
    }
    
    /// Get a transaction receipt.
    /// - Parameter hash: Transaction hash.
    /// - Returns: Receipt data.
    public func getTransactionReceipt(hash: String) async throws -> [String: Any] {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_getTransactionReceipt", params: [hash])
        return response as? [String: Any] ?? [:]
    }
    
    /// Get the current block number.
    /// - Returns: Block number (decimal).
    public func getBlockNumber() async throws -> Int {
        guard !rpcUrl.isEmpty else {
            throw EVMError.rpcNotConfigured
        }
        let response = try await rpcCall(method: "eth_blockNumber")
        if let hex = response as? String {
            return Int(hex.replacingOccurrences(of: "0x", with: ""), radix: 16) ?? 0
        }
        return 0
    }
    
    // MARK: - Chain Lookup
    
    /// Find a chain by ID.
    /// - Parameter chainId: Chain ID to find.
    /// - Returns: Chain configuration, or nil.
    public func findChain(chainId: Int) -> ChainConfig? {
        chains.first { $0.chainId == chainId }
    }
    
    // MARK: - Formatting
    
    /// Format a transaction into a JSON-RPC compatible format.
    /// - Parameter tx: Transaction request.
    /// - Returns: Dictionary for JSON-RPC params.
    public static func formatTransaction(_ tx: TransactionRequest) -> [String: String] {
        var formatted: [String: String] = [:]
        if !tx.from.isEmpty { formatted["from"] = tx.from }
        if !tx.to.isEmpty { formatted["to"] = tx.to }
        if let value = tx.value { formatted["value"] = value }
        if let data = tx.data { formatted["data"] = data }
        if let gas = tx.gas { formatted["gas"] = gas }
        if let gasPrice = tx.gasPrice { formatted["gasPrice"] = gasPrice }
        if let maxFee = tx.maxFeePerGas { formatted["maxFeePerGas"] = maxFee }
        if let maxPriority = tx.maxPriorityFeePerGas { formatted["maxPriorityFeePerGas"] = maxPriority }
        if let nonce = tx.nonce { formatted["nonce"] = nonce }
        if let chainId = tx.chainId { formatted["chainId"] = String(format: "0x%x", chainId) }
        return formatted
    }
    
    /// Convert a hex balance string to ETH.
    /// - Parameter hex: Hex balance.
    /// - Returns: ETH as a decimal string.
    public static func hexToEth(_ hex: String) -> String {
        let cleanHex = hex.replacingOccurrences(of: "0x", with: "")
        guard let wei = UInt64(cleanHex, radix: 16) else { return "0" }
        let eth = Double(wei) / 1e18
        return String(format: "%.6f", eth)
    }
    
    // MARK: - Private
    
    private func buildCallParams(_ params: EthCallParams) -> [String: Any] {
        var dict: [String: Any] = [:]
        dict["to"] = params.to
        if let data = params.data { dict["data"] = data }
        if let from = params.from { dict["from"] = from }
        if let value = params.value { dict["value"] = value }
        return dict
    }
    
    private func rpcCall(method: String, params: [Any] = []) async throws -> Any {
        let body: [String: Any] = [
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1
        ]
        
        guard let url = URL(string: rpcUrl) else {
            throw EVMError.invalidRpcUrl(rpcUrl)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw EVMError.rpcError("HTTP error")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EVMError.rpcError("Invalid JSON response")
        }
        
        if let error = json["error"] as? [String: Any],
           let message = error["message"] as? String {
            throw EVMError.rpcError(message)
        }
        
        return json["result"] ?? ""
    }
}

// MARK: - Errors

/// EVM adapter errors.
public enum EVMError: Error, LocalizedError {
    case rpcNotConfigured
    case invalidRpcUrl(String)
    case rpcError(String)
    
    public var errorDescription: String? {
        switch self {
        case .rpcNotConfigured:
            return "RPC URL not configured. Call setRpcUrl() first."
        case let .invalidRpcUrl(url):
            return "Invalid RPC URL: \(url)"
        case let .rpcError(msg):
            return "RPC error: \(msg)"
        }
    }
}

File: onux/packages/android-kotlin/src/main/kotlin/com/onchainux/chain/EvmAdapter.kt
/**
 * EvmAdapter — EVM chain adapter for Android.
 *
 * Provides EVM-specific operations via JSON-RPC:
 * - Balance fetching (eth_getBalance)
 * - Gas estimation (eth_estimateGas)
 * - Contract calls (eth_call)
 * - Block number, nonce, gas price queries
 * - Transaction receipt retrieval
 *
 * ## Usage
 * ```kotlin
 * val adapter = EvmAdapter("https://eth.llamarpc.com")
 * val balance = adapter.getBalance("0x...")
 * val blockNumber = adapter.getBlockNumber()
 * ```
 */
package com.cinacoin.chain

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import java.net.HttpURLConnection
import java.net.URL

// ────────────────────────────────────────────────────────────────────────────
// RPC Types
// ────────────────────────────────────────────────────────────────────────────

@Serializable
data class EvmRpcRequest(
    val jsonrpc: String = "2.0",
    val method: String,
    val params: JsonArray,
    val id: Int = 1
)

@Serializable
data class EvmRpcResponse(
    val jsonrpc: String? = null,
    val result: JsonElement? = null,
    val error: RpcError? = null,
    val id: Int? = null
)

// ────────────────────────────────────────────────────────────────────────────
// Transaction Request for EVM
// ────────────────────────────────────────────────────────────────────────────

data class EvmTransactionRequest(
    val from: String,
    val to: String,
    val value: String? = null,
    val data: String? = null,
    val gas: String? = null,
    val gasPrice: String? = null,
    val maxFeePerGas: String? = null,
    val maxPriorityFeePerGas: String? = null,
    val nonce: String? = null,
    val chainId: Int? = null
) {
    fun toJson(): JsonObject = buildJsonObject {
        put("from", JsonPrimitive(from))
        put("to", JsonPrimitive(to))
        value?.let { put("value", JsonPrimitive(it)) }
        data?.let { put("data", JsonPrimitive(it)) }
        gas?.let { put("gas", JsonPrimitive(it)) }
        gasPrice?.let { put("gasPrice", JsonPrimitive(it)) }
        maxFeePerGas?.let { put("maxFeePerGas", JsonPrimitive(it)) }
        maxPriorityFeePerGas?.let { put("maxPriorityFeePerGas", JsonPrimitive(it)) }
        nonce?.let { put("nonce", JsonPrimitive(it)) }
        chainId?.let { put("chainId", JsonPrimitive("0x${it.toString(16)}")) }
    }
}

// ────────────────────────────────────────────────────────────────────────────
// EVM Chain Config
// ────────────────────────────────────────────────────────────────────────────

data class EvmChainConfig(
    val chainId: Int,
    val name: String,
    val rpcUrl: String,
    val nativeCurrency: EvmNativeCurrency,
    val blockExplorerUrl: String? = null,
    val testnet: Boolean = false
) {
    companion object {
        val ethereum = EvmChainConfig(
            chainId = 1, name = "Ethereum", rpcUrl = "https://eth.llamarpc.com",
            nativeCurrency = EvmNativeCurrency(name = "Ether", symbol = "ETH", decimals = 18),
            blockExplorerUrl = "https://etherscan.io"
        )
        val polygon = EvmChainConfig(
            chainId = 137, name = "Polygon", rpcUrl = "https://polygon-rpc.com",
            nativeCurrency = EvmNativeCurrency(name = "MATIC", symbol = "MATIC", decimals = 18),
            blockExplorerUrl = "https://polygonscan.com"
        )
        val arbitrum = EvmChainConfig(
            chainId = 42161, name = "Arbitrum", rpcUrl = "https://arb1.arbitrum.io/rpc",
            nativeCurrency = EvmNativeCurrency(name = "Ether", symbol = "ETH", decimals = 18),
            blockExplorerUrl = "https://arbiscan.io"
        )
        val optimism = EvmChainConfig(
            chainId = 10, name = "Optimism", rpcUrl = "https://mainnet.optimism.io",
            nativeCurrency = EvmNativeCurrency(name = "Ether", symbol = "ETH", decimals = 18),
            blockExplorerUrl = "https://optimistic.etherscan.io"
        )
        val base = EvmChainConfig(
            chainId = 8453, name = "Base", rpcUrl = "https://mainnet.base.org",
            nativeCurrency = EvmNativeCurrency(name = "Ether", symbol = "ETH", decimals = 18),
            blockExplorerUrl = "https://basescan.org"
        )
    }
}

data class EvmNativeCurrency(
    val name: String,
    val symbol: String,
    val decimals: Int
)

// ────────────────────────────────────────────────────────────────────────────
// EVM Adapter
// ────────────────────────────────────────────────────────────────────────────

/**
 * Real EVM JSON-RPC adapter.
 */
class EvmAdapter(
    private var rpcUrl: String = "https://eth.llamarpc.com",
    private val httpClient: EvmHttpClient = DefaultEvmHttpClient()
) {

    private var _connectedAddress: String? = null

    /** Connected EVM wallet address. */
    val connectedAddress: String?
        get() = _connectedAddress

    /** RPC endpoint. */
    var endpoint: String
        get() = rpcUrl
        set(value) { rpcUrl = value }

    // ─── Presets ────────────────────────────────────────────────────────

    companion object {
        val MAINNET = "https://eth.llamarpc.com"
        val POLYGON = "https://polygon-rpc.com"
        val ARBITRUM = "https://arb1.arbitrum.io/rpc"
        val OPTIMISM = "https://mainnet.optimism.io"
        val BASE = "https://mainnet.base.org"

        fun isValidAddress(address: String): Boolean {
            return address.length == 42 && address.startsWith("0x") &&
                address.substring(2).all { it.isDigit() || it in 'a'..'f' || it in 'A'..'F' }
        }

        /** Convert wei to ETH string. */
        fun weiToEth(wei: Long): String = String.format("%.6f", wei / 1e18)

        /** Convert ETH to wei. */
        fun ethToWei(eth: Double): Long = (eth * 1e18).toLong()

        /** Parse hex string to BigInteger. */
        fun hexToBigInteger(hex: String): BigInteger {
            val cleaned = if (hex.startsWith("0x")) hex.substring(2) else hex
            return BigInteger(cleaned, 16)
        }

        /** Convert BigInteger to hex string. */
        fun bigIntToHex(value: BigInteger): String = "0x" + value.toString(16)
    }

    // ─── Balance ────────────────────────────────────────────────────────

    /**
     * Get ETH balance for an address via `eth_getBalance`.
     * @return Balance in ETH as decimal string (e.g. "1.234567").
     */
    suspend fun getBalance(address: String, blockTag: String = "latest"): String =
        withContext(Dispatchers.IO) {
            require(isValidAddress(address)) { "Invalid EVM address: $address" }

            val result = rpcCall("eth_getBalance", listOf(
                JsonPrimitive(address),
                JsonPrimitive(blockTag)
            ))
            val balanceHex = result.jsonPrimitive.content
            val wei = hexToBigInteger(balanceHex)
            val eth = wei.toBigDecimal().divide(
                BigDecimal("1000000000000000000"), 6, java.math.RoundingMode.HALF_UP
            )
            eth.toPlainString()
        }

    /**
     * Get raw balance in wei (hex string).
     */
    suspend fun getBalanceRaw(address: String, blockTag: String = "latest"): String =
        withContext(Dispatchers.IO) {
            require(isValidAddress(address)) { "Invalid EVM address: $address" }
            val result = rpcCall("eth_getBalance", listOf(
                JsonPrimitive(address), JsonPrimitive(blockTag)
            ))
            result.jsonPrimitive.content
        }

    // ─── Gas ────────────────────────────────────────────────────────────

    /**
     * Get the current gas price via `eth_gasPrice`.
     * @return Gas price in wei (hex string).
     */
    suspend fun getGasPrice(): String = withContext(Dispatchers.IO) {
        val result = rpcCall("eth_gasPrice", emptyList())
        result.jsonPrimitive.content
    }

    /**
     * Estimate gas for a transaction via `eth_estimateGas`.
     * @return Gas estimate (hex string).
     */
    suspend fun estimateGas(tx: EvmTransactionRequest): String =
        withContext(Dispatchers.IO) {
            val result = rpcCall("eth_estimateGas", listOf(tx.toJson()))
            result.jsonPrimitive.content
        }

    /**
     * Get max priority fee per gas via `eth_maxPriorityFeePerGas`.
     * @return Max priority fee (hex string).
     */
    suspend fun getMaxPriorityFeePerGas(): String = withContext(Dispatchers.IO) {
        val result = rpcCall("eth_maxPriorityFeePerGas", emptyList())
        result.jsonPrimitive.content
    }

    // ─── Block Number ──────────────────────────────────────────────────

    /**
     * Get the current block number via `eth_blockNumber`.
     * @return Block number as Long.
     */
    suspend fun getBlockNumber(): Long = withContext(Dispatchers.IO) {
        val result = rpcCall("eth_blockNumber", emptyList())
        hexToBigInteger(result.jsonPrimitive.content).toLong()
    }

    // ─── Chain ID ──────────────────────────────────────────────────────

    /**
     * Get the chain ID via `eth_chainId`.
     * @return Chain ID as Int.
     */
    suspend fun getChainId(): Int = withContext(Dispatchers.IO) {
        val result = rpcCall("eth_chainId", emptyList())
        hexToBigInteger(result.jsonPrimitive.content).toInt()
    }

    // ─── Nonce ─────────────────────────────────────────────────────────

    /**
     * Get the nonce for an address via `eth_getTransactionCount`.
     * @return Nonce as Long.
     */
    suspend fun getNonce(address: String, blockTag: String = "pending"): Long =
        withContext(Dispatchers.IO) {
            require(isValidAddress(address)) { "Invalid EVM address: $address" }
            val result = rpcCall("eth_getTransactionCount", listOf(
                JsonPrimitive(address), JsonPrimitive(blockTag)
            ))
            hexToBigInteger(result.jsonPrimitive.content).toLong()
        }

    // ─── Contract Calls ────────────────────────────────────────────────

    /**
     * Call a contract read method via `eth_call`.
     * @param to Contract address.
     * @param data Encoded function call data.
     * @param blockTag Block to query (default: "latest").
     * @return Result data (hex string).
     */
    suspend fun call(
        to: String,
        data: String,
        from: String? = null,
        blockTag: String = "latest"
    ): String = withContext(Dispatchers.IO) {
        val params = buildJsonObject {
            put("to", JsonPrimitive(to))
            put("data", JsonPrimitive(data))
            from?.let { put("from", JsonPrimitive(it)) }
        }
        val result = rpcCall("eth_call", listOf(params, JsonPrimitive(blockTag)))
        result.jsonPrimitive.content
    }

    /**
     * Get ERC-20 token balance for an address.
     * @param tokenAddress ERC-20 contract address.
     * @param userAddress User wallet address.
     * @return Token balance (hex string).
     */
    suspend fun getTokenBalance(tokenAddress: String, userAddress: String): String =
        withContext(Dispatchers.IO) {
            require(isValidAddress(tokenAddress)) { "Invalid token address" }
            require(isValidAddress(userAddress)) { "Invalid user address" }

            val paddedAddress = userAddress.substring(2).padStart(64, '0')
            val data = "0x70a08231$paddedAddress"

            call(to = tokenAddress, data = data)
        }

    // ─── Transactions ──────────────────────────────────────────────────

    /**
     * Get a transaction by hash via `eth_getTransactionByHash`.
     * @return Transaction data as JSON element.
     */
    suspend fun getTransaction(txHash: String): JsonElement = withContext(Dispatchers.IO) {
        rpcCall("eth_getTransactionByHash", listOf(JsonPrimitive(txHash)))
    }

    /**
     * Get a transaction receipt via `eth_getTransactionReceipt`.
     * @return Receipt data as JSON element.
     */
    suspend fun getTransactionReceipt(txHash: String): JsonElement = withContext(Dispatchers.IO) {
        rpcCall("eth_getTransactionReceipt", listOf(JsonPrimitive(txHash)))
    }

    /**
     * Send a raw signed transaction via `eth_sendRawTransaction`.
     * @param rawTx Hex-encoded signed transaction.
     * @return Transaction hash.
     */
    suspend fun sendRawTransaction(rawTx: String): String = withContext(Dispatchers.IO) {
        val result = rpcCall("eth_sendRawTransaction", listOf(JsonPrimitive(rawTx)))
        result.jsonPrimitive.content
    }

    // ─── Connected Address ─────────────────────────────────────────────

    /**
     * Set the connected wallet address (called after successful connection).
     */
    fun setConnectedAddress(address: String) {
        require(isValidAddress(address)) { "Invalid EVM address: $address" }
        _connectedAddress = address
    }

    /**
     * Disconnect the current wallet.
     */
    fun disconnect() {
        _connectedAddress = null
    }

    // ─── EIP-1193 Compatible Request ───────────────────────────────────

    /**
     * EIP-1193 compatible request method for EVM operations.
     */
    suspend fun request(method: String, params: List<JsonElement> = emptyList()): JsonElement =
        when (method) {
            "eth_getBalance" -> JsonPrimitive(getBalance(
                params.firstOrNull()?.jsonPrimitive?.content ?: "",
                params.getOrNull(1)?.jsonPrimitive?.content ?: "latest"
            ))
            "eth_gasPrice" -> JsonPrimitive(getGasPrice())
            "eth_estimateGas" -> {
                // Would need to parse params into EvmTransactionRequest
                JsonPrimitive(estimateGas(EvmTransactionRequest(
                    from = "", to = ""
                )))
            }
            "eth_blockNumber" -> JsonPrimitive(getBlockNumber().toString())
            "eth_chainId" -> JsonPrimitive(getChainId().toString(16))
            "eth_getTransactionCount" -> JsonPrimitive(getNonce(
                params.firstOrNull()?.jsonPrimitive?.content ?: "",
                params.getOrNull(1)?.jsonPrimitive?.content ?: "pending"
            ).toString())
            "eth_call" -> {
                val callParams = params.firstOrNull()?.jsonObject ?: JsonObject(emptyMap())
                call(
                    to = callParams["to"]?.jsonPrimitive?.content ?: "",
                    data = callParams["data"]?.jsonPrimitive?.content ?: "0x",
                    blockTag = params.getOrNull(1)?.jsonPrimitive?.content ?: "latest"
                ).let { JsonPrimitive(it) }
            }
            "eth_sendRawTransaction" -> JsonPrimitive(sendRawTransaction(
                params.firstOrNull()?.jsonPrimitive?.content ?: ""
            ))
            else -> throw EvmError.UnsupportedMethod(method)
        }

    // ─── Private: RPC ───────────────────────────────────────────────────

    private suspend fun rpcCall(method: String, params: List<JsonElement>): JsonElement =
        withContext(Dispatchers.IO) {
            val request = EvmRpcRequest(
                method = method,
                params = JsonArray(params)
            )
            val body = kotlinx.serialization.json.Json.encodeToString(request)
            val responseBody = httpClient.post(rpcUrl, body)

            val response = kotlinx.serialization.json.Json.decodeFromString<EvmRpcResponse>(responseBody)

            if (response.error != null) {
                throw EvmError.RpcError(
                    "[${response.error.code}] ${response.error.message}"
                )
            }

            response.result ?: JsonPrimitive("")
        }
}

// ────────────────────────────────────────────────────────────────────────────
// Errors
// ────────────────────────────────────────────────────────────────────────────

/** EVM adapter errors. */
sealed class EvmError : Exception() {
    data class InvalidAddress(val address: String) : EvmError()
    object NotConnected : EvmError()
    data class InvalidRpcUrl(val url: String) : EvmError()
    data class RpcError(val message: String) : EvmError()
    object InvalidParams : EvmError()
    data class UnsupportedMethod(val method: String) : EvmError()
    data class NotImplemented(val message: String) : EvmError()

    override val message: String
        get() = when (this) {
            is InvalidAddress -> "Invalid EVM address: $address"
            NotConnected -> "Not connected to an EVM wallet"
            is InvalidRpcUrl -> "Invalid RPC URL: $url"
            is RpcError -> "RPC error: $message"
            InvalidParams -> "Invalid parameters"
            is UnsupportedMethod -> "Unsupported method: $method"
            is NotImplemented -> "Not implemented: $message"
        }
}

// ────────────────────────────────────────────────────────────────────────────
// HTTP Client Interface
// ────────────────────────────────────────────────────────────────────────────

/** HTTP client interface for EVM RPC calls. */
interface EvmHttpClient {
    suspend fun post(url: String, body: String): String
}

/** Default implementation using Java HttpURLConnection. */
class DefaultEvmHttpClient : EvmHttpClient {
    override suspend fun post(url: String, body: String): String =
        withContext(Dispatchers.IO) {
            val connection = URL(url).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                connection.doOutput = true
                connection.connectTimeout = 15_000
                connection.readTimeout = 30_000

                connection.outputStream.write(body.toByteArray(Charsets.UTF_8))
                connection.outputStream.flush()

                val responseCode = connection.responseCode
                if (responseCode != HttpURLConnection.HTTP_OK) {
                    val errorBody = connection.errorStream?.bufferedReader()?.use { it.readText() }
                        ?: "HTTP $responseCode"
                    throw EvmError.RpcError("HTTP $responseCode: $errorBody")
                }

                connection.inputStream.bufferedReader().use { it.readText() }
            } finally {
                connection.disconnect()
            }
        }
}
